In [198]:
import pandas as pd

### Review the unstructured csv files and answer the following questions with code that supports your conclusions:

Are there any data quality issues present?

Are there any fields that are challenging to understand?

In [199]:
product_path = 'C:/Users/ericp/Downloads/Fetch_Take_Home/PRODUCTS_TAKEHOME.csv'
transaction_path = 'C:/Users/ericp/Downloads/Fetch_Take_Home/TRANSACTION_TAKEHOME.csv'
user_path = 'C:/Users/ericp/Downloads/Fetch_Take_Home/USER_TAKEHOME.csv'

In [200]:
# Read csv using pandas
product = pd.read_csv(product_path, dtype={'BARCODE': str}) # assigning data type for BARCODE
transaction = pd.read_csv(transaction_path, dtype={'BARCODE': str})
user = pd.read_csv(user_path)

## Product Table

In [253]:
product.head(5)

CATEGORY_1              CATEGORY_2                   CATEGORY_3  \
0  Health & Wellness           Sexual Health  Conductivity Gels & Lotions   
1             Snacks           Puffed Snacks         Cheese Curls & Puffs   
2  Health & Wellness               Hair Care        Hair Care Accessories   
3  Health & Wellness               Oral Care                   Toothpaste   
4  Health & Wellness  Medicines & Treatments               Essential Oils   

  CATEGORY_4                                       MANUFACTURER  \
0        NaN                                                NaN   
1        NaN                                                NaN   
2        NaN                           PLACEHOLDER MANUFACTURER   
3        NaN                                  COLGATE-PALMOLIVE   
4        NaN  MAPLE HOLISTICS AND HONEYDEW PRODUCTS INTERCHA...   

             BRAND        BARCODE  
0              NaN   796494407820  
1              NaN   023278011028  
2          ELECSOP   461817824225  
3          COLGATE   035000466815  
4  MAPLE HOLISTICS  0806810850459

In [254]:
# Check for missing data
missing_data_percentage = product.isnull().sum() / product.shape[0] * 100

# Find duplicate rows
duplicate_rows = product.duplicated().sum()

# Check data types
data_types = product.dtypes

print("Missing Data (%):\n", missing_data_percentage, '\n')
print("Duplicate Rows:\n", duplicate_rows, '\n')
print("Data Types:\n", data_types)

Missing Data (%):
 CATEGORY_1       0.013128
CATEGORY_2       0.168411
CATEGORY_3       7.162895
CATEGORY_4      92.021898
MANUFACTURER    26.784160
BRAND           26.783923
BARCODE          0.476020
dtype: float64 

Duplicate Rows:
 57 

Data Types:
 CATEGORY_1      object
CATEGORY_2      object
CATEGORY_3      object
CATEGORY_4      object
MANUFACTURER    object
BRAND           object
BARCODE         object
dtype: object


In [255]:
# Take a look at those records with no value missing
product[~product['CATEGORY_4'].isnull()][:5]

CATEGORY_1     CATEGORY_2              CATEGORY_3  \
15  Health & Wellness      Hair Care              Hair Tools   
25  Health & Wellness   Hair Removal     Shaving Gel & Cream   
31  Health & Wellness      Skin Care  Lip Balms & Treatments   
39             Snacks  Puffed Snacks                 Popcorn   
44  Health & Wellness   Hair Removal     Shaving Gel & Cream   

                     CATEGORY_4              MANUFACTURER               BRAND  \
15         Hair Brushes & Combs  PLACEHOLDER MANUFACTURER           BULK BUYS   
25  Women's Shaving Gel & Cream  PLACEHOLDER MANUFACTURER                WAHL   
31                    Lip Balms        THE CLOROX COMPANY  BURT'S BEES BEAUTY   
39       Already Popped Popcorn      STONEHEDGE FARMS INC    STONEHEDGE FARMS   
44  Women's Shaving Gel & Cream                CONAIR LLC              CONAIR   

         BARCODE  
15  731015145751  
25  043917102238  
31  792850662094  
39  045764333544  
44  191566227113

### Summary
There are several issues in this dataset **PRODUCT** :

1. Unclear Column Names:

- The columns CATEGORY_1 to CATEGORY_4 have ambiguous names. Based on the data, these columns could represent:
    - CATEGORY_1 -> Department
    - CATEGORY_2 to CATEGORY_4 -> Category / Subcategory / Product Type

2. Missing Data:
- There is a significant amount of missing data across the dataset, particularly in CATEGORY_4, where 92% of the values are missing.
- Approximately 27% of BRAND & MANUFACTURER data is missing, which could cause inaccuracy in related analysis.
3. Column *BARCODE*:
- The BARCODE column, which should ideally contain a unique identifier for each product, has missing values. This could compromise its use as a primary key for product identification.

## Transaction Table

In [256]:
transaction.head()

RECEIPT_ID PURCHASE_DATE  \
0  0000d256-4041-4a3e-adc4-5623fb6e0c99    2024-08-21   
1  0001455d-7a92-4a7b-a1d2-c747af1c8fd3    2024-07-20   
2  00017e0a-7851-42fb-bfab-0baa96e23586    2024-08-18   
3  000239aa-3478-453d-801e-66a82e39c8af    2024-06-18   
4  00026b4c-dfe8-49dd-b026-4c2f0fd5c6a1    2024-07-04   

                   SCAN_DATE STORE_NAME                   USER_ID  \
0  2024-08-21 14:19:06.539 Z    WALMART  63b73a7f3d310dceeabd4758   
1  2024-07-20 09:50:24.206 Z       ALDI  62c08877baa38d1a1f6c211a   
2  2024-08-19 15:38:56.813 Z    WALMART  60842f207ac8b7729e472020   
3  2024-06-19 11:03:37.468 Z  FOOD LION  63fcd7cea4f8442c3386b589   
4  2024-07-05 15:56:43.549 Z   RANDALLS  6193231ae9b3d75037b0f928   

        BARCODE FINAL_QUANTITY FINAL_SALE  
0  015300014978           1.00             
1           NaN           zero       1.49  
2  078742229751           1.00             
3  783399746536           zero       3.49  
4  047900501183           1.00

In [257]:
# Check for missing data
missing_data_percentage = transaction.isnull().sum() / transaction.shape[0] * 100

# Find duplicate rows
duplicate_rows = transaction.duplicated().sum()

# Check data types
data_types = transaction.dtypes

print("Missing Data (%):\n", missing_data_percentage, '\n')
print("Duplicate Rows:\n", duplicate_rows, '\n')
print("Data Types:\n", data_types)

Missing Data (%):
 RECEIPT_ID         0.000
PURCHASE_DATE      0.000
SCAN_DATE          0.000
STORE_NAME         0.000
USER_ID            0.000
BARCODE           11.524
FINAL_QUANTITY     0.000
FINAL_SALE         0.000
dtype: float64 

Duplicate Rows:
 171 

Data Types:
 RECEIPT_ID        object
PURCHASE_DATE     object
SCAN_DATE         object
STORE_NAME        object
USER_ID           object
BARCODE           object
FINAL_QUANTITY    object
FINAL_SALE        object
dtype: object


While there're only a few duplicated rows, I noticed that there're a lot more duplicates in "RECEIPT_ID" .

In [258]:
# Check duplicates for "RECEIPT_ID"
transaction[['RECEIPT_ID']].duplicated().sum()

25560

In [259]:
# Take a look at those duplicated records
transaction[transaction['RECEIPT_ID'].duplicated(keep=False)].sort_values(['RECEIPT_ID'])[:10]

RECEIPT_ID PURCHASE_DATE  \
0      0000d256-4041-4a3e-adc4-5623fb6e0c99    2024-08-21   
41567  0000d256-4041-4a3e-adc4-5623fb6e0c99    2024-08-21   
1      0001455d-7a92-4a7b-a1d2-c747af1c8fd3    2024-07-20   
39291  0001455d-7a92-4a7b-a1d2-c747af1c8fd3    2024-07-20   
2      00017e0a-7851-42fb-bfab-0baa96e23586    2024-08-18   
25928  00017e0a-7851-42fb-bfab-0baa96e23586    2024-08-18   
3      000239aa-3478-453d-801e-66a82e39c8af    2024-06-18   
41475  000239aa-3478-453d-801e-66a82e39c8af    2024-06-18   
43233  00026b4c-dfe8-49dd-b026-4c2f0fd5c6a1    2024-07-04   
4      00026b4c-dfe8-49dd-b026-4c2f0fd5c6a1    2024-07-04   

                       SCAN_DATE STORE_NAME                   USER_ID  \
0      2024-08-21 14:19:06.539 Z    WALMART  63b73a7f3d310dceeabd4758   
41567  2024-08-21 14:19:06.539 Z    WALMART  63b73a7f3d310dceeabd4758   
1      2024-07-20 09:50:24.206 Z       ALDI  62c08877baa38d1a1f6c211a   
39291  2024-07-20 09:50:24.206 Z       ALDI  62c08877baa38d1a1f6c211a   
2      2024-08-19 15:38:56.813 Z    WALMART  60842f207ac8b7729e472020   
25928  2024-08-19 15:38:56.813 Z    WALMART  60842f207ac8b7729e472020   
3      2024-06-19 11:03:37.468 Z  FOOD LION  63fcd7cea4f8442c3386b589   
41475  2024-06-19 11:03:37.468 Z  FOOD LION  63fcd7cea4f8442c3386b589   
43233  2024-07-05 15:56:43.549 Z   RANDALLS  6193231ae9b3d75037b0f928   
4      2024-07-05 15:56:43.549 Z   RANDALLS  6193231ae9b3d75037b0f928   

            BARCODE FINAL_QUANTITY FINAL_SALE  
0      015300014978           1.00             
41567  015300014978           1.00       1.54  
1               NaN           zero       1.49  
39291           NaN           1.00       1.49  
2      078742229751           1.00             
25928  078742229751           1.00       2.54  
3      783399746536           zero       3.49  
41475  783399746536           1.00       3.49  
43233  047900501183           1.00       5.29  
4      047900501183           1.00

It appears that the duplications in the RECEIPT_ID column have nearly identical data, with differences observed only in the "FINAL_QUANTITY" and "FINAL_SALE" columns.

To investigate further, identify the "RECEIPT_ID" that occurs most frequently.

In [260]:
# Count the frequency of each 'RECEIPT_ID' and get the top 3
transaction['RECEIPT_ID'].value_counts()[:3]

RECEIPT_ID
bedac253-2256-461b-96af-267748e6cecf    12
bc304cd7-8353-4142-ac7f-f3ccec720cb3     8
4ec870d2-c39f-4a40-bf8a-26a079409b20     8
Name: count, dtype: int64

In [261]:
# Inspect the most freqenty appeared 'RECEIPT_ID'
transaction[transaction['RECEIPT_ID'] == 'bedac253-2256-461b-96af-267748e6cecf']

RECEIPT_ID PURCHASE_DATE  \
18637  bedac253-2256-461b-96af-267748e6cecf    2024-09-08   
18638  bedac253-2256-461b-96af-267748e6cecf    2024-09-08   
18639  bedac253-2256-461b-96af-267748e6cecf    2024-09-08   
18640  bedac253-2256-461b-96af-267748e6cecf    2024-09-08   
18641  bedac253-2256-461b-96af-267748e6cecf    2024-09-08   
18642  bedac253-2256-461b-96af-267748e6cecf    2024-09-08   
39703  bedac253-2256-461b-96af-267748e6cecf    2024-09-08   
39704  bedac253-2256-461b-96af-267748e6cecf    2024-09-08   
39705  bedac253-2256-461b-96af-267748e6cecf    2024-09-08   
39706  bedac253-2256-461b-96af-267748e6cecf    2024-09-08   
39707  bedac253-2256-461b-96af-267748e6cecf    2024-09-08   
39708  bedac253-2256-461b-96af-267748e6cecf    2024-09-08   

                       SCAN_DATE STORE_NAME                   USER_ID  \
18637  2024-09-08 20:00:42.348 Z     KROGER  614f7e8081627974a57c8a9e   
18638  2024-09-08 20:00:42.348 Z     KROGER  614f7e8081627974a57c8a9e   
18639  2024-09-08 20:00:42.348 Z     KROGER  614f7e8081627974a57c8a9e   
18640  2024-09-08 20:00:42.348 Z     KROGER  614f7e8081627974a57c8a9e   
18641  2024-09-08 20:00:42.348 Z     KROGER  614f7e8081627974a57c8a9e   
18642  2024-09-08 20:00:42.348 Z     KROGER  614f7e8081627974a57c8a9e   
39703  2024-09-08 20:00:42.348 Z     KROGER  614f7e8081627974a57c8a9e   
39704  2024-09-08 20:00:42.348 Z     KROGER  614f7e8081627974a57c8a9e   
39705  2024-09-08 20:00:42.348 Z     KROGER  614f7e8081627974a57c8a9e   
39706  2024-09-08 20:00:42.348 Z     KROGER  614f7e8081627974a57c8a9e   
39707  2024-09-08 20:00:42.348 Z     KROGER  614f7e8081627974a57c8a9e   
39708  2024-09-08 20:00:42.348 Z     KROGER  614f7e8081627974a57c8a9e   

            BARCODE FINAL_QUANTITY FINAL_SALE  
18637  011110121141           zero       0.89  
18638  011110121141           1.00             
18639  011110121141           zero       0.89  
18640  011110121141           1.00             
18641  011110121141           zero       0.89  
18642  011110121141           1.00             
39703  011110121141           1.00       0.89  
39704  011110121141           1.00       0.89  
39705  011110121141           1.00       0.89  
39706  011110121141           1.00       0.89  
39707  011110121141           1.00       0.89  
39708  011110121141           1.00       0.89

We can observe that all 12 records are nearly identical, with the only differences being the occurrences of "zero" and blank values.

Based on this, it is likely that these records represent the same data. The discrepancies might stem from issues in the data entry process or other underlying problems.

### Summary
There are several issues in the **TRANSACTION** dataset:

1. Column *BARCODE*:
- Approximately 11.5% of the data in the BARCODE column is null.
2. Column *FINAL_QUANTITY*:
- The values in the FINAL_QUANTITY column are inconsistent, with 0 written as "zero" in some cases while other numbers are written numerically.
- Clarification Needed: It seems illogical for the quantity to be 0 but still appear on a receipt. Is "zero" used as a placeholder, or is it a data entry issue?
3. Column *FINAL_SALE*:
- Around 12,500 records (approximately one-fourth of the dataset) have blank values in the FINAL_SALE column.
4. Column *RECEIPT_ID*:
- While RECEIPT_ID appears to be a unique identifier for each receipt, there are numerous duplications.
- Clarification Needed: Upon investigation, these duplications might represent the same records. Further explanation and validation are required.

## User Table

In [262]:
user.head()

ID               CREATED_DATE  \
0  5ef3b4f17053ab141787697d  2020-06-24 20:17:54.000 Z   
1  5ff220d383fcfc12622b96bc  2021-01-03 19:53:55.000 Z   
2  6477950aa55bb77a0e27ee10  2023-05-31 18:42:18.000 Z   
3  658a306e99b40f103b63ccf8  2023-12-26 01:46:22.000 Z   
4  653cf5d6a225ea102b7ecdc2  2023-10-28 11:51:50.000 Z   

                  BIRTH_DATE STATE LANGUAGE  GENDER  
0  2000-08-11 00:00:00.000 Z    CA   es-419  female  
1  2001-09-24 04:00:00.000 Z    PA       en  female  
2  1994-10-28 00:00:00.000 Z    FL   es-419  female  
3                        NaN    NC       en     NaN  
4  1972-03-19 00:00:00.000 Z    PA       en  female

In [263]:
# Check for missing data
missing_data_percentage = user.isnull().sum() / user.shape[0] * 100

# Find duplicate rows
duplicate_rows = user.duplicated().sum()

# Check data types
data_types = user.dtypes

print("Missing Data (%):\n", missing_data_percentage, '\n')
print("Duplicate Rows:\n", duplicate_rows, '\n')
print("Data Types:\n", data_types)

Missing Data (%):
 ID               0.000
CREATED_DATE     0.000
BIRTH_DATE       3.675
STATE            4.812
LANGUAGE        30.508
GENDER           5.892
dtype: float64 

Duplicate Rows:
 0 

Data Types:
 ID              object
CREATED_DATE    object
BIRTH_DATE      object
STATE           object
LANGUAGE        object
GENDER          object
dtype: object


In [264]:
# Check duplications for unique ID
user.duplicated(['ID']).sum()

0

In [265]:
# Check number of unique value for each string column
string_columns = user.select_dtypes(include=['object'])
string_summary = {col: user[col].nunique() for col in string_columns}
print(string_summary)

{'ID': 100000, 'CREATED_DATE': 99942, 'BIRTH_DATE': 54721, 'STATE': 52, 'LANGUAGE': 2, 'GENDER': 11}


### Summary

There are several issues identified in the **USER** dataset:

1. Missing Data:

- Significant missing data in the LANGUAGE column (30.508%), with smaller missing percentages in BIRTH_DATE, STATE, and GENDER. These gaps could impact demographic analysis.

2. Duplicate Rows:
- No duplicate rows were detected, which is a positive sign.
3. Column *CREATED_DATE*:

- The presence of the Z in the timestamp format may require extra steps before querying, should be reformatted before being entered into the database.
- The timestamp includes seconds and milliseconds, which may not be necessary depending on the specific use case.
4. Column *BIRTH_DATE*:

- The inclusion of time (00:00:00.000 Z) in the date is irrelevant for birth dates and could be simplified before analysis.

____________________________________

## Joined Table
After looking separately into the datasets, now I also want to check on the joined dataset.

In [266]:
# Join TRANSACTION and PRODUCT table together, using left join to find out if products in all transaction can be found in PRODUCT table
transaction_product = pd.merge(transaction[transaction['BARCODE'].notnull()], product, on='BARCODE', how='left')

In [267]:
transaction_product.head()

RECEIPT_ID PURCHASE_DATE  \
0  0000d256-4041-4a3e-adc4-5623fb6e0c99    2024-08-21   
1  00017e0a-7851-42fb-bfab-0baa96e23586    2024-08-18   
2  000239aa-3478-453d-801e-66a82e39c8af    2024-06-18   
3  00026b4c-dfe8-49dd-b026-4c2f0fd5c6a1    2024-07-04   
4  0002d8cd-1701-4cdd-a524-b70402e2dbc0    2024-06-24   

                   SCAN_DATE STORE_NAME                   USER_ID  \
0  2024-08-21 14:19:06.539 Z    WALMART  63b73a7f3d310dceeabd4758   
1  2024-08-19 15:38:56.813 Z    WALMART  60842f207ac8b7729e472020   
2  2024-06-19 11:03:37.468 Z  FOOD LION  63fcd7cea4f8442c3386b589   
3  2024-07-05 15:56:43.549 Z   RANDALLS  6193231ae9b3d75037b0f928   
4  2024-06-24 19:44:54.247 Z    WALMART  5dcc6c510040a012b8e76924   

        BARCODE FINAL_QUANTITY FINAL_SALE CATEGORY_1              CATEGORY_2  \
0  015300014978           1.00                Pantry  Packaged Meals & Sides   
1  078742229751           1.00                   NaN                     NaN   
2  783399746536           zero       3.49        NaN                     NaN   
3  047900501183           1.00                   NaN                     NaN   
4  681131411295           zero       1.46     Snacks             Snack Cakes   

                     CATEGORY_3                     CATEGORY_4  \
0            Microwavable Sides  Microwavable Non-Potato Sides   
1                           NaN                            NaN   
2                           NaN                            NaN   
3                           NaN                            NaN   
4  Cakes & Truffles Snack Cakes                            NaN   

               MANUFACTURER          BRAND  
0                   PEPSICO        CHEETOS  
1                       NaN            NaN  
2                       NaN            NaN  
3                       NaN            NaN  
4  PLACEHOLDER MANUFACTURER  PRIVATE LABEL

In [268]:
# Check the number of records that all data from PRODUCT table are null
transaction_product.iloc[:, 8:].isnull().all(axis=1).sum() / transaction_product.shape[0]

0.4388082643880826

### Summary

It appears that over 43% of the products in the transaction data cannot be directly matched with the product table. This discrepancy could lead to issues in analyses involving product sales and may require further investigation to address the mismatches.

In [269]:
# Join TRANSACTION and USER table together, using left join to find out if users in all transaction can be found in USER table
transaction_user =pd.merge(transaction[transaction['USER_ID'].notnull()], user, left_on='USER_ID', right_on='ID', how='left')

In [246]:
# Check the number of records that all data from User table are null
transaction_user.iloc[:, 8:].isnull().all(axis=1).sum() / transaction_user.shape[0]

0.99476

### Summary

It appears that over over 99% of the users in the transaction data cannot be directly matched with the user table. This discrepancy could lead to issues in analyses involving user data and may require further investigation to address the mismatches.

In [248]:
transaction_product_inner = pd.merge(transaction, user, left_on='USER_ID', right_on='ID', how='inner')
transaction_product_user_inner = pd.merge(transaction_product_inner, product, on='BARCODE', how='inner')
